feat(router-lite): navigation strategy

Sayan751 opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting since someone might have pushed the same thing before!
-->

# Pull Request

## 📖 Description

<!---
Provide some background and a description of your work.
-->

This is a port of navigation strategy of au1 router. An example looks as follows.


```ts
import { customElement } from '@aurelia/runtime-html';
import { route, NavigationStrategy, RecognizedRoute } from '@aurelia/router-lite';
import template from './my-app.html';
import { About } from './about';
import { Home } from './home';

let dataLoaded = false;

@route({
  routes: [
    {
      path: ['', 'home'],
      component: Home,
    },
    {
      path: 'about',
      component: About,
    },
    {
      path: 'foo',
      component: new NavigationStrategy((viewportInstruction: IViewportInstruction, ctx: IRouteContext, node: RouteNode, route: RecognizedRoute<unknown>) => {
        if (dataLoaded) return Home;

        dataLoaded = true;
        return About;
      })
    }
  ],
})
export class MyApp {}
```

cc: @simonfox 

### 🎫 Issues

<!---
* List and link relevant issues here.
-->

## 👩‍💻 Reviewer Notes

<!---
Provide some notes for reviewers to help them provide targeted feedback.
-->

## 📑 Test Plan

<!---
Please provide a summary of the tests affected by this work and any unique strategies employed in testing the features/fixes.
-->

## ⏭ Next Steps

<!---
If there is relevant follow-up work to this PR, please list any existing issues or provide brief descriptions of what you would like to do next.
-->

<!--
Love Aurelia? Please consider supporting our collective:
👉  https://opencollective.com/aurelia
-->

simonfox commented:
getting this when linking my repro project to your branch
![image](https://github.com/user-attachments/assets/c69bd3a5-64bd-4559-bf6e-1061a027367a)

all I've changed is package.json reference to local build of your branch
![image](https://github.com/user-attachments/assets/aebc29fa-9db5-4247-af44-c354ee93ddf0)

simonfox commented:
@Sayan751 possibly an issue with parameterized routes. Trying to sort a minimal repro for you.

simonfox commented:
> @Sayan751 possibly an issue with parameterized routes. Trying to sort a minimal repro for you.

ok so maybe not an _issue_, but if I have the following route config that uses a navigation strategy
```
{
          id: "feature-load-id",
          path: ":id",
          title: "Load",
          component: new NavigationStrategy((viewportInstruction: IViewportInstruction, ctx: IRouteContext, node: RouteNode) => {
            console.log("nav strategy: ", viewportInstruction.params, ctx);
            if (!this._loaded) {
              this._loaded = true;
              return FeatureLoading;
            } else {
              return FeatureLoad;
            }
          }),
          nav: false,
        },
```
I want to be able to get the value of the id param so I can determine if that id is loaded. `viewportInstruction.params` must be params of the parent?

bigopon commented on packages/router-lite/src/route.ts:
```diff
@@ -21,9 +22,14 @@ export class RouteConfig implements IRouteConfig, IChildRouteConfig {
+    const ceDfn = CustomElement.getDefinition(this._component as RouteType);
     return this._path = [ceDfn.name, ...ceDfn.aliases];
   }
+  private _currentComponent: Routeable | null = null;
```
```suggestion
  /** @internal */
  private _currentComponent: Routeable | null = null;
```

bigopon commented on packages/router-lite/src/route.ts:
```diff
@@ -36,14 +42,25 @@ export class RouteConfig implements IRouteConfig, IChildRouteConfig {
+  ) {
+    // eslint-disable-next-line no-cond-assign
+    if (this.isNavigationStrategy = _component instanceof NavigationStrategy) {
+      navigatingObserver?.subscribe({
+        handleChange: (newValue, _previousValue) => {
+          if (newValue === true) this._currentComponent = null;
+        }
+      });
```
don't we need to unsubscribe?

Sayan751 replied:
The route config objects have a lifetime similar to the application lifetime. Thus, a disposal of those objects are of little use. I can explain this by annotating the code with comments. However, if you think that unsubscribe is needed, let me know, I can surely arrange that.

bigopon commented on packages/router-lite/src/route.ts:
```diff
@@ -36,14 +42,25 @@ export class RouteConfig implements IRouteConfig, IChildRouteConfig {
+    private readonly _component: Routeable | NavigationStrategy,
     public readonly nav: boolean,
-  ) { }
+    private readonly navigatingObserver: IObserver | null,
```
`navigatingObserver` seems disconnected with a `RouteConfig`, Rather than making `RouteConfig` know about some foreign concepts/ideas, what about we do the subscription somewhere else and let `RouteConfig` expose a method/property to mutate the current component instead?

Sayan751 replied:
Refactored the infra.

bigopon reviewed (commented):
Nice work @Sayan751 only a few small comments.

Sayan751 commented:
@simonfox For easily handling the parameters, I have added a fourth parameter to the navigation strategy factory function. Also added a test as an example. You may have a look if the example in the test case covers your use-case.

simonfox commented:
Thanks @Sayan751! I'm getting the `rtInvalidOperationNavigationStrategyComponent` error in our app when I try to redirect back using `router.load`. However I've been able to do what I want to when trying to repro. I'll keep trying to get to the bottom of it, but confirm that the extra param does what I need it to.

simonfox commented:
@Sayan751 this is not the exact issue I'm getting in our app, but could be related - https://github.com/simonfox/router-issue

A child router doesn't seem to get activated after using nav strategy. See `src/feature/load`. Nav strategy is used to choose between `feature-loading` and `feature-loaded`. When redirecting to `feature-loaded` after doing the loading from `feature-loading` the child router activation of loaded doesn't happen i.e. getRouteConfig isnt called and the au-viewport doesn't get registered.

Note that I've used dumber deps property (in `gulpfile.js`) to link to a local build of aurelia. You may want to just drop the example into your own test app.

bigopon reviewed (approved):
nice work @Sayan751 , though I think we probably should have private props prefixed with `_`, there's no reason to delay this small optimization.

Sayan751 commented:
@simonfox I have added another test for your scenario. It seems to be working. I would recommend calling the `router.load` in the `attached` hook.

I will go ahead and merge this PR.

simonfox commented:
@Sayan751  Scenario I was trying to describe would mean `C12` component would describe a child router and template contain a `au-viewport`

In that case the get routes callback wasn't being invoked.

Sayan751 commented:
@simonfox Alright. I will refactor the test.

Sayan751 commented:
@simonfox Seems to work as well.
